src/Codegen/TypeDef.hs

```diff
@@ -1,10 +1,87 @@
+{-# LANGUAGE QuasiQuotes #-}
 module Codegen.TypeDef where
 
 import qualified IR.Types.Flat                 as L
 import qualified IR.Types.TypeSystem           as L
 
+import           Language.C.Quote.GCC
 import qualified Language.C.Syntax             as C
 
+import           Common.Identifiers             ( DConId
+                                                , TConId
+                                                , fromString
+                                                , ident
+                                                )
+
+import           Codegen.Identifiers
+import           Data.Char                      ( toLower )
+import           Data.List                      ( maximumBy
+                                                , partition
+                                                )
+
 -- | Generate definitions for SSM type definitions.
-genTypeDef :: L.TypeDef L.Type -> C.Definition
-genTypeDef = error "Not yet implemented"
+genTypeDef :: TConId -> L.TypeDef L.Type -> (C.Definition, C.Definition)
+genTypeDef tconid (L.TypeDef dCons _) = (tags, structDef)
+ where
+  structDef = [cedecl| typedef struct {
+                       $ty:ssm_mm_md* $id:header;
+                       $ty:ssm_object_t* $id:payload[($ty:word_t ) $uint:maxNumFields]; 
+                     } $id:tconid;     
+                   |]
+  tags = [cedecl| enum $id:enumName {
+                          $enums:(enumElts)
+                       };  
+             |]
+
+  -- | find any data constructors that hold less than 32 bits of data
+  (big, small ) = partition (\(_, sz) -> sz >= 32) $ augmentWSize <$> dCons
+  -- | find any data constructors that are < 32 bits when including their tag bits
+  (_  , medium) = smallEnough small
+  -- | arrange data constructors from smallest to largest tag value
+  tagged        = small ++ medium ++ big
+
+  -- | define parts of the c definitions
+  enumName      = CIdent $ fromString $ lower $ ident tconid
+  enumElts =
+    [cenum|$id:(fst $ fst $ head tagged)|]
+      : (toCEnum . fst . fst <$> tail tagged)
+  maxNumFields =
+    snd $ maximumBy (\(_, sz1) (_, sz2) -> sz1 `compare` sz2) (big ++ medium)
+
+  -- | helper functions
+
+  -- | turn DConId into an elt of a c enumerated type
+  toCEnum :: DConId -> C.CEnum
+  toCEnum n = [cenum|$id:n|]
+
+  -- | categorize data constructor as either small enough or too big to be represented in 31 bits
+  -- | when taking into account the tag bits
+  smallEnough
+    :: [((DConId, L.TypeVariant L.Type), Int)]
+    -> ( [((DConId, L.TypeVariant L.Type), Int)]
+       , [((DConId, L.TypeVariant L.Type), Int)]
+       )
+  smallEnough candidates = partition (\(_, sz) -> (sz + tagBits) < 32)
```

Comment:
this can be rewritten as ```partition ((< 32) . (+ tagBits) . dConSize) candidates```

Reply:
better yet, introuce a named helper under `tagBits` to make it clear what you're partition on.